compiler/hash-ast/src/resolve.rs

```diff
@@ -0,0 +1,122 @@
+//! Hash Compiler module resolving logic
+//!
+//! All rights reserved 2021 (c) The Hash Language authors
+
+use crate::{
+    error::ParseResult,
+    fs::resolve_path,
+    location::SourceLocation,
+    module::ModuleIdx,
+    parse::{ParserBackend, ParsingContext},
+};
+use derive_getters::Getters;
+use derive_more::Constructor;
+use hash_utils::timed;
+use log::{debug, Level};
+use rayon::Scope;
+use std::{fs, path::Path};
+
+pub trait ModuleResolver {
+    fn module_source(&self) -> Option<&str>;
+    fn add_module(
+        &mut self,
+        import_path: impl AsRef<Path>,
+        location: Option<SourceLocation>,
+    ) -> ParseResult<ModuleIdx>;
+}
+
+#[derive(Debug, Copy, Clone, Constructor, Getters)]
```

Comment:
What is `Getters`, are they just derived functions to get a specific field, why is it better over just getting the field?

Reply:
Yes. Better because when dealing with value types, they prevent you from modifying the field, and allow only read access through a reference. Technically not necessary, but without them I'd also have to make the struct fields public.

Reply:
Well, it seems to me that one way or another, they'll have to be 'public'. I just feel like it's unnecessary bloat that Java introduces to create 'encapsulation'.

Reply:
To be a bit pragmatic, I actually think there is some merit to this 'encapsulation' pattern, since it restricts the API of the type, which if not done might lead to unwanted consequences elsewhere in the codebase. However, since these types are localised to this crate, I think they would be fine just being plain structs.